### PR TITLE
Add qgsVariantEqual for null-aware comparison

### PR DESCRIPTION
**Pull request by elpaso**
As requested in https://github.com/qgis/QGIS/pull/8079

### CODE REVIEW

File: src/core/qgis.h [hunk 1]
```diff
@@ -448,6 +448,17 @@ CORE_EXPORT qlonglong qgsPermissiveToLongLong( QString string, bool &ok );
  */
 CORE_EXPORT bool qgsVariantLessThan( const QVariant &lhs, const QVariant &rhs );
 
+/**
+ * Compares two QVariant values and return whether they are equal, it takes into
+ * account null values.
+ *
+ * \param lhs first value
+ * @param rhs second value
```
Comment (m-kuhn) [inline src/core/qgis.h]: `\param` or `@para,`?
If in doubt, use `\a` ;)

File: src/core/qgis.h [hunk 2]
```diff
@@ -448,6 +448,17 @@ CORE_EXPORT qlonglong qgsPermissiveToLongLong( QString string, bool &ok );
  */
 CORE_EXPORT bool qgsVariantLessThan( const QVariant &lhs, const QVariant &rhs );
 
+/**
+ * Compares two QVariant values and return whether they are equal, it takes into
+ * account null values.
```
Comment (m-kuhn) [inline src/core/qgis.h]: Better be explicit here, like "NULL values are treated as equal".

File: tests/src/core/testqgis.cpp [hunk 3]
```diff
@@ -359,6 +360,21 @@ void TestQgis::testQgsRound()
...
+  QVariant rhs;
+
+  QVERIFY( lhs == rhs );
+  lhs.setValue( 0 );
+  QVERIFY( lhs != rhs );
+  rhs.setValue( 0 );
+  QVERIFY( lhs == rhs );
+  lhs.setValue( 1.2345 );
+  rhs.setValue( 1.2345 );
+  QVERIFY( lhs == rhs );
```
Comment (m-kuhn) [inline tests/src/core/testqgis.cpp]: Doest it work with `==`?

In any case, also add null variants (currently it's only testing invalid ones.

```
lhs = QVariant( QVariant::Int );
QVERIFY( !qgsVariantEqual( lhs, rhs ) );

rhs = QVariant( QVariant::Int );
QVERIFY( qgsVariantEqual( lhs, rhs ) )

```

**Comment by m-kuhn** [comment]
:+1:

File: tests/src/core/testqgis.cpp [hunk 4]
```diff
@@ -359,6 +360,36 @@ void TestQgis::testQgsRound()
...
+  QVERIFY( qgsVariantEqual( QVariant(), QVariant() ) );
+  QVERIFY( QVariant() == QVariant() );
+
+  // Zero
+  QVERIFY( qgsVariantEqual( QVariant( 0 ), QVariant( 0.0f ) ) );
+  QVERIFY( QVariant( 0 ) == QVariant( 0.0f ) );
+
+  // Double
+  QVERIFY( qgsVariantEqual( QVariant( 1.234 ), QVariant( 1.234 ) ) );
+  QVERIFY( qgsVariantEqual( QVariant( QVariant::Int ), QVariant( QVariant::Int ) ) );
```
Comment (m-kuhn) [inline tests/src/core/testqgis.cpp]: That's already tested a couple of lines down I think

File: tests/src/core/testqgis.cpp [hunk 5]
```diff
@@ -359,6 +360,36 @@ void TestQgis::testQgsRound()
   QGSCOMPARENEAR( qgsRound( 9999999.87654321987654321, 14 ), 9999999.876543219876543, 0.00000000000001 );
 }
 
+void TestQgis::testQgsVariantEqual()
+{
+
+  // Invalid
+  QVERIFY( qgsVariantEqual( QVariant(), QVariant() ) );
+  QVERIFY( QVariant() == QVariant() );
```
Comment (m-kuhn) [inline tests/src/core/testqgis.cpp]: Just wondering where you want to go with those checks... Trigger a warning when Qt upstream introduces changes to `QVariant::operator==`?

Comment (elpaso) [reply tests/src/core/testqgis.cpp]: no, it's just to confirm that I'm only changing the comparison with NULL vs 0

**Review by m-kuhn — APPROVED** [review]

**Comment by elpaso** [comment]
@m-kuhn thanks for the patient review :)

I'll merge as soon as Travis is happy.

**Comment by m-kuhn** [comment]
Thank you @elpaso this was much needed.

If you want to use it in some places:

* src/core/qgis.cpp:192
* src/core/qgsfeaturefiltermodel.cpp:536
* src/gui/qgsattributeform.cpp:360
* src/providers/postgres/qgspostgresprovider.cpp:2121

**Pull request merged**